File: problems_list/1372.二叉树中的最长交错路径/1372-二叉树中的最长交错路径.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestZigZag = function(root) {
    let ans=0;
    function dfs(root, left, len){
        if(!root) return;
        ans=Math.max(ans, len);
        if(left){
            dfs(root.left, true, 1);
            dfs(root.right, false, len+1);
        }else{
            dfs(root.left, true, len+1);
            dfs(root.right, false, 1);
        }
    }
    dfs(root, true, 0);
    return ans;
};

File: problems_list/1372.二叉树中的最长交错路径/1372-二叉树中的最长交错路径.py
class Solution:
    def longestZigZag(self, root: TreeNode) -> int:
        def dfs(root: TreeNode, p_left: bool, l: int):
            nonlocal ans
            if not root:
                return
            ans=max(ans,l)
            if p_left:
                dfs(root.left,True,1)
                dfs(root.right,False,l+1)
            else:
                dfs(root.left,True,l+1)
                dfs(root.right,False,1)
            
        ans=0
        dfs(root,True,0)
        return ans